src/generators/notebookgenerator/itemrenderer.tsx

```diff
@@ -13,8 +15,161 @@ import { sanitizerOptions } from '../shared';
+  if (match) {
+    return parseInt(match[1], 10);
+  }
+  return -1;
+}
+
+/**
+ * Collapses a notebook cell by hiding its section-defined sub-cells (i.e., cells which have lower precedence).
+ *
+ * @private
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function collapseCell(tracker: any, cell: Cell): void {
+  // Guard against attempting to collapse already hidden cells...
+  if (cell.isHidden) {
+    return;
+  }
+  const level: number = headerLevel(cell);
+
+  // Guard against attempting to collapse cells which are not "collapsible" (i.e., do not define sections)...
+  if (level === -1) {
+    return;
+  }
+  const widgets = tracker.currentWidget.content.widgets;
+  const len = widgets.length;
+  const idx = widgets.indexOf(cell);
+
+  // Guard against an unrecognized "cell" argument...
+  if (idx === -1) {
+    return;
+  }
+  // Search for notebook cells which are semantically defined as sub-cells...
+  for (let i = idx + 1; i < len; i++) {
+    let w = widgets[i];
+    let l: number = headerLevel(w);
+
+    // Check if a widget is at the same or higher level...
+    if (l >= 0 && l <= level) {
+      // We've reached the end of the section...
+      break;
+    }
+    // Collapse a sub-cell by setting its `hidden` state:
+    w.setHidden(true);
+  }
+  // Set a meta-data flag to indicate that we've collapsed notebook sections:
+  cell.model.metadata.set('toc-nb-collapsed', true);
+}
+
+/**
+ * Expands a notebook cell by displaying its section-defined sub-cells (i.e., cells which have lower precedence).
+ *
+ * @private
+ * @param tracker - notebook tracker
+ * @param cell - notebook cell
+ */
+function uncollapseCell(tracker: any, cell: Cell): void {
+  // Guard against attempting to un-collapse cells which we did not collapse or are already un-collapsed...
+  if (
+    cell.model.metadata.has('toc-nb-collapsed') === false ||
+    cell.model.metadata.get('toc-nb-collapsed') === false
+  ) {
+    return;
+  }
+  const level: number = headerLevel(cell);
+
+  // Guard against attempting to un-collapse cells which are not "collapsible" (i.e., do not define sections)...
+  if (level === -1) {
+    return;
+  }
+  const widgets = tracker.currentWidget.content.widgets;
+  const len = widgets.length;
+  const idx = widgets.indexOf(cell);
+
+  // Guard against an unrecognized "cell" argument...
+  if (idx === -1) {
+    return;
+  }
+  // Search for notebook cells which are semantically defined as sub-cells...
+  for (let i = idx + 1; i < len; i++) {
+    let w = widgets[i];
+    let l: number = headerLevel(w);
+
+    // Check if a widget is at the same or higher level...
+    if (l >= 0 && l <= level) {
+      // We've reached the end of the section...
+      break;
+    }
+    // Un-collapse a sub-cell by setting its `hidden` state:
+    w.setHidden(false);
+  }
+  // Set a meta-data flag to indicate that we've un-collapsed notebook sections:
+  cell.model.metadata.set('toc-nb-collapsed', false);
```

Comment:
I'm still not convinced about the choice to use metadata here (which was not your choice, @kgryte), but as long as we are using it -- I think we should remove the metadata here rather than setting to `false`.

Reply:
Addressed. Deleting a field may have a negative effect on performance, but likely to not be rate-limiting given everything else which is happening when un-collapsing a section.